provider/aws: Add AWS RDS Read Replica

catsby opened the pull request:
My proposal for adding support to for read replicas to `resource_aws_db`, by way of adding a `replicate_source_db` attribute, referencing the `identifier` of the source database.

**Caveats**:

- Replicas inherit many of their attributes from the primary database (`username`, `password`, `engine`). Many of these are required when creating a database instance, but not when creating a replica. The trouble there is if being a replica is just a flag on a resource, that resource still needs those required attributes, even if they're ignored (inherited) from the source. Until Terraform has a `ValidateFunc` or `RequiredIf` concept(s), we're stuck repeating ourselves. 

- Replicas can be promoted to full, standalone databases. Terraform can support this by sending a `PromoteReadReplica` API call when the `replicate_source_db` attribute is removed in the configuration. This may be a surprise to some users, since we can't notify them that "removing this attribute makes a full read/write database" during `plan`. 

- If you use the example config below and let TF do the inheritance of attributes, you may get in a weird state when you promote. Manual managing of the config file will be required if you need to separate the two database more. 

**Example config:**

```javascript
resource "aws_db_instance" "primary" {
  identifier = "tf-primary-db"
  allocated_storage = 5
  engine = "mysql"
  engine_version = "5.6.19b"
  instance_class = "db.t1.micro"
  name = "baz"
  password = "fofofofxx"
  username = "foo"
  multi_az = false
  apply_immediately = true
  backup_retention_period = 1
  tags {
    Name = "tf-primary-db"
  }
  parameter_group_name = "default.mysql5.6"
}

resource "aws_db_instance" "replica" {
  identifier = "tf-replica-db"
  replicate_source_db = "${aws_db_instance.primary.identifier}"
  allocated_storage = "${aws_db_instance.primary.allocated_storage}"
  engine = "${aws_db_instance.primary.engine}"
  engine_version = "${aws_db_instance.primary.engine_version}"
  instance_class = "${aws_db_instance.primary.instance_class}"
  password = "${aws_db_instance.primary.password}"
  username = "${aws_db_instance.primary.username}"
  tags {
    Name = "tf-replica-db"
  }
}
```

In practice, users may want to manually duplicate all those fields. In the implementation, we'll have to ignore most of them that don't apply. 

The fields that are applicable to replica creation:

```go
AutoMinorVersionUpgrade *bool 
AvailabilityZone *string
DBInstanceClass *string
DBInstanceIdentifier *string
DBSubnetGroupName *string
IOPS *int64
OptionGroupName *string
Port *int64
PubliclyAccessible *bool
SourceDBInstanceIdentifier *string
StorageType *string
Tags []*Tag
```

**Relevant docs:**

- [DB Instance Replication](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.Replication.html)
- [Working with PostgreSQL and MySQL Read Replicas](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_ReadRepl.html#USER_ReadRepl.Promote)
- [API Docs for `create`](http://docs.aws.amazon.com/AmazonRDS/latest/APIReference//API_CreateDBInstanceReadReplica.html)

**Things:**

- [x] Docs (mostly)
- [x] Example config
- [x] Implementation 
- [x] Acceptance tests


I welcome thoughts on this approach

catsby commented:
Implementation is (mostly) done. You can create, destroy, and promote a replica. To promote, simply remove the `replicate_source_db` attribute in the config, that will send a `PromoteReadReplica` API call and sever the replication link. 

The acceptance test is incomplete, but I wanted to see if anyone had feedback on implementation / usage.

catsby commented:
cc @radeksimko, @ryanking as at least 2 people who I know requested this in some fashion :smile:

ryanking commented:
Interface lgtm.

radeksimko commented:
Overall looks good, I have not been looking at the API docs yet, but since we gonna be able to build a M-S-S RDS cluster, it would be nice to test typical scenarios, like engine upgrade.

I know that RDS does some clever things like "upgrade each slave separately first", then upgrade master, while promoting one of the slaves to master = no downtime caused. I'm not sure if all this happens under the hood automatically, I haven't tried that yet, but it's a typical scenario that should work well.

catsby commented:
> I know that RDS does some clever things like "upgrade each slave separately first", then upgrade master, while promoting one of the slaves to master = no downtime caused. 

Wow, is this process documented somewhere? I didn't read the docs entirely so I may have just missed it. Seems like promoting a follower would be a bad thing, since it has different address. Unless you're talking about their HA setup (`multi_az`), which is a bit different; they'll swap out DNS internally in that case. 

What are you thinking I need to do here to accommodate?

radeksimko commented:
> What are you thinking I need to do here to accommodate?

Like I mentioned, I did not actually experienced this, I've just read it somewhere, can't remember where, sorry...

What I was rather trying to suggest is to battle test this in these situations, when you upgrade certain things. Just make sure that the most basic expectable upgrade paths are working.

catsby commented:
Rebased on master and added an acceptance test.
I apologize in advance for time spent running that acceptance test... creating primary, backup primary, create replica, test things, _then_ destroy primary, destroy replica ... it takes a long time.

@phinze please review when you get a chance.

radeksimko commented on website/source/docs/providers/aws/r/db_instance.html.markdown:
```diff
@@ -65,6 +66,17 @@ The following arguments are supported:
+ more information on using Replication.
+
+
+~> **NOTE:** Removing the `relicate_source` attribute from an existing RDS
```
Oops, typo here - `relicate_source` => `replicate_source`

radeksimko commented on website/source/docs/providers/aws/r/db_instance.html.markdown:
```diff
@@ -48,7 +48,8 @@ The following arguments are supported:
 * `availability_zone` - (Optional) The AZ for the RDS instance.
-* `backup_retention_period` - (Optional) The days to retain backups for.
+* `backup_retention_period` - (Optional) The days to retain backups for. Must be
+`1` or greater to be a source for a [Read Replicate][1].
```
Typo or maybe just a naming convention inconsistence? `Read Replicate` => `Read Replica`

radeksimko commented on builtin/providers/aws/resource_aws_db_instance_test.go:
```diff
@@ -156,3 +187,38 @@ resource "aws_db_instance" "bar" {
+	resource "aws_db_instance" "replica" {
+	  identifier = "tf-replica-db-%d"
+		backup_retention_period = 0
+	  replicate_source_db = "${aws_db_instance.bar.identifier}"
```
Indentation here.

radeksimko commented on builtin/providers/aws/resource_aws_db_instance_test.go:
```diff
@@ -49,6 +49,26 @@ func TestAccAWSDBInstance(t *testing.T) {
+		CheckDestroy: testAccCheckAWSDBInstanceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccReplicaInstanceConfig(rand.New(rand.NewSource(time.Now().UnixNano())).Int()),
```
Can't we use the unique ID generator that Paul made recently https://github.com/hashicorp/terraform/pull/1632 ?

radeksimko replied:
Actually, forget it, probably not a good idea to make tests depending on more things...

radeksimko commented on builtin/providers/aws/resource_aws_db_instance.go:
```diff
@@ -397,6 +435,23 @@ func resourceAwsDbInstanceRead(d *schema.ResourceData, meta interface{}) error {
+	if v.ReadReplicaSourceDBInstanceIdentifier != nil {
+		log.Printf("\n\n------\nread replica instance identifier: %#v", *v.ReadReplicaSourceDBInstanceIdentifier)
+	} else {
+		log.Printf("\n\n------\nno replica identifier")
```
Do we still need these `\n\n------\n` here?

catsby replied:
haha, no, that's my debugging statements :laughing:  fixed in 6de8f9d

radeksimko commented:
Overall LGTM. :+1: 

```
make testacc TEST=github.com/hashicorp/terraform/builtin/providers/aws TESTARGS='-run=DBInstance' 2>/dev/null
go generate ./...
TF_ACC=1 go test github.com/hashicorp/terraform/builtin/providers/aws -v -run=DBInstance -timeout 45m
=== RUN TestAccAWSDBInstance
--- PASS: TestAccAWSDBInstance (618.04s)
=== RUN TestAccAWSDBInstanceReplica
--- PASS: TestAccAWSDBInstanceReplica (1620.66s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	2238.710s
```

yet I haven't actually tried building a replicated RDS outside of the test suite.

This resource should make decent amount of people happier. :smile:

phinze commented on website/source/docs/providers/aws/r/db_instance.html.markdown:
```diff
@@ -65,6 +66,17 @@ The following arguments are supported:
 * `apply_immediately` - (Optional) Specifies whether any database modifications
      are applied immediately, or during the next maintenance window. Default is
      `false`. See [Amazon RDS Documentation for more for more information.](http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Overview.DBInstance.Modifying.html)
+* `replicate_source` - (Optional) Specifies that this resource is a Replicate
```
It's `replicate_source_db` in code and `replicate_source` here.

phinze commented:
One wee docs thing, but otherwise LGTM!
